src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/TypeOperations.cs

```diff
@@ -2552,101 +2545,35 @@ protected internal override Expression VisitRuntimeVariables(RuntimeVariablesExp
 
             private static IRuntimeVariables MergeRuntimeVariables(IRuntimeVariables first, IRuntimeVariables second, int[] indexes)
             {
-                return new MergedRuntimeVariables(first, second, indexes);
+                return new RuntimeOps.MergedRuntimeVariables(first, second, indexes);
             }
 
             protected internal override Expression VisitParameter(ParameterExpression node)
             {
-                LocalVariable var;
-                if (_variables.TryGetValue(node, out var))
+                IStrongBox box = GetBox(node);
+                if (box == null)
                 {
-                    return Expression.Convert(
-                        Expression.Field(
-                            var.LoadFromArray(
-                                Expression.Constant(_frame.Data),
-                                Expression.Constant(_frame.Closure),
-                                node.Type
-                            ),
-                            "Value"
-                        ),
-                        node.Type
-                    );
+                    return node;
                 }
-                return node;
+                return Expression.Convert(Expression.Field(Expression.Constant(box), "Value"), node.Type);
             }
 
-            private sealed class RuntimeVariables : IRuntimeVariables
+            private IStrongBox GetBox(ParameterExpression variable)
```

Comment:
All difference between compiler and interpreter implementation of the quoter are now:

* Constructor and fields
* `GetBox` implementation
* Absence/presence of `Convert` in `VisitConstant`
* Use of `Call` versus `Invoke` in `VisitRuntimeVariables` (due to issue with late bound calls on AOT?)

It seems we'll be able to share most of the code through a base class with at least one virtual for `GetBox`. If the two `Visit` methods can't converge easily, both can override a virtual to construct the replacement node in the way they like.